### ISSUE TITLE
bigquery: consider moving from getQueryResults to listTableData

### ISSUE THREAD

**Issue by stephenplusplus** [googleapis/nodejs-bigquery#12]
_From @pongad on October 31, 2017 3:35_

The current `query` method uses [`getQueryResults`](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/getQueryResults) RPC to read rows of the results. However, benchmark shows that [`listTableData`](https://cloud.google.com/bigquery/docs/reference/rest/v2/tabledata/list) run significantly faster. Experiments in Go shows `getQueryResults` take 70% longer.

The choice of RPC is an implementation detail and can be changed whenever. However, `listTableData` returns fewer fields. In Java, we decided to tweak the API surface to make it easier to change RPC later. Of course, the change to API needs to happen before GA.

Will Node require any API change?

_Copied from original issue: GoogleCloudPlatform/google-cloud-node#2709_

**Comment by stephenplusplus** [comment]
Could you explain the RPC calls in order from "SELECT url FROM [publicdata:samples.github_nested] LIMIT 100" to "resultingRows.forEach(...)"?

**Comment by stephenplusplus** [comment]
_From @pongad on October 31, 2017 22:28_

Current Go implementation works like this:
1. use `job.insert` to start the job
2. use `getQueryResults` with `max_results=0` to wait for the job to finish
  a. you can read the schema off of this RPC
3. use `tabledata.list` to actually iterate over results.

Currently in Java, we return some statistics with *each page*. These stats will be removed because `tabledata.list` doesn't return these stats.

**Comment by stephenplusplus** [comment]
_From @pongad on November 8, 2017 3:40_

I believe this should be P1. If this requires surface changes, it must happen before GA.

**Comment by stephenplusplus** [comment]
_From @callmehiphop on November 8, 2017 16:30_

I feel like I'm definitely missing part of the picture here, at first glance it seems like `listTableData` doesn't have an equivalent for advanced queries where sorting/filtering is required.

Node currently supports the `listTableData` rpc via [bigquery/table#getRows](https://googlecloudplatform.github.io/google-cloud-node/#/docs/google-cloud/0.56.0/bigquery/table?method=getRows) - if there are perf concerns, could we not just recommend users check out that method for simpler queries?

**Comment by stephenplusplus** [comment]
_From @callmehiphop on November 8, 2017 18:16_

Nevermind! I just looked through the Go implementation and connected all the dots. I do not believe a surface change will be necessary in order to swap.

**Comment by stephenplusplus** [comment]
_From @pongad on November 9, 2017 0:29_

@callmehiphop Thank you for verifying. Since no surface change is required for swap, this is no longer a GA blocker. I'll put this at P2, but please change if you feel differently.

**Comment by stephenplusplus** [comment]
@callmehiphop has a PR ready (#484), but blocked it based on [this comment](https://github.com/googleapis/nodejs-bigquery/pull/484#issuecomment-518339740):

> Technically this will be a TypeScript breaking change, so I'm going to hold off on merging until a major bump presents itself.

**Comment by meredithslota** [comment]
We have an independent effort that is targeting improving query latency so we are closing this in favor of that.
